src/llvm_ts.cpp

```diff
@@ -287,10 +286,31 @@ static void submit_cpu(KernelType type, Func &&func, uint32_t width,
     jitc_task = new_task;
 }
 
-Task *LLVMThreadState::launch(Kernel kernel, uint32_t size,
-                              std::vector<void *> *kernel_params,
-                              uint32_t /* kernel_param_count */,
-                              const uint8_t* /* kernel_params_global */) {
+/// Temporary scratch space for scheduled tasks
+static std::vector<Task *> scheduled_tasks;
+
+void LLVMThreadState::barrier() {
+    if (scheduled_tasks.size() == 1) {
+        task_release(jitc_task);
+        jitc_task = scheduled_tasks[0];
+    } else {
+        jitc_assert(!scheduled_tasks.empty(),
+                    "jit_eval(): no tasks generated!");
+
+        // Insert a barrier task
+        Task *new_task = task_submit_dep(nullptr, scheduled_tasks.data(),
+                                         (uint32_t) scheduled_tasks.size());
+        task_release(jitc_task);
+        for (Task *t : scheduled_tasks)
+            task_release(t);
+        jitc_task = new_task;
+    }
+    scheduled_tasks.clear();
+}
+
+Task *LLVMThreadState::launch(Kernel kernel, KernelKey *, XXH128_hash_t,
```

Comment:
I like the convention of putting the param name in an inline comment if it's unused, though I don't know if that's used consistently in this codebase:
```cpp
Kernel kernel, KernelKey * /*key*/, ...
```